<center> <h2> DS 3000 - Summer 2020</h2> </center>
<center> <h3> DS Report </h3> </center>


<center> <h3> Analysis and Prediction of Twitch Channel Growth</h3> </center>
<center><h4>Carrine Shrestha, Yihang Li</h4></center>


<hr style="height:2px; border:none; color:black; background-color:black;">

#### Executive Summary:

In the current world, streaming has become a popular form of entertainment and as with any source of entertainment, there are entertainment providers. Many entertainment providers look at the service as a job and expect a source of revenue. We analyzed the use of the streaming service, Twitch, and the predictability of income growth among top streamers. Income was measured based on revenue gained through subscriptions. 
We scraped data from TwitchTracker to gain information on subscriptions of channels and exported data off of SullyGnome to gain insight on features to use as analysis. We hypothesized that followers gained would be the most influential feature in predicting income growth. We were able to gain insight on our data through the use of scatterplots, histograms, and violin plots. Machine learning algorithms for regression were applied to the dataset. However, our findings were shown as being insignificant through ANOVA analysis. 

<hr style="height:2px; border:none; color:black; background-color:black;">

## Outline
1. <a href='#1'>INTRODUCTION</a>
2. <a href='#2'>METHOD</a>
3. <a href='#3'>RESULTS</a>
4. <a href='#4'>DISCUSSION</a>

<a id="1"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 1. INTRODUCTION

**Problem Statement**

With the growth of live streaming as an entertainment platform, coronavirus has presented a strong opportunity for online streaming platforms to expand and grow. We want to track this growth and how growth is displayed among the creators in Twitch. Understanding Twitch’s growth and live streaming as a career field is important to those looking into streaming as a viable at-home, virtual job, and predicting measures of growth and success. We will create an expected projection of Twitch channels over the next year on a monthly basis of viewership and other metrics, which will be used in parallel with subscriber data to project expected income growth. We also want to gain insight into the jump in viewership due to coronavirus and better understand the rise of the streaming industry. 

**Significance of the Problem**

Live streaming is an industry that is quickly growing in the technology space. Following Twitch’s initial success in the early 2010s, Youtube (via Youtube Live), Facebook (Facebook/Instagram Live) and other major platforms have followed up by creating their own streaming platforms for content creators. Insights into Twitch’s growth potential can quantify the growth of live streaming and their platforms as a brand new career field with opportunities for people to look into. Tracking the projected growth of a fast-moving industry in live streaming can be really useful for choosing a job as a software or data engineer, as joining one of these companies can be particularly lucrative. Alternatively, should investment opportunities present themselves in the future, having an understanding of YoY growth and metric-backed expectations can lead to potentially rewarding investments. Finally, if someone were to come across the expected growth and data and were interested in becoming a livestreamer themselves, they could use the projected data to measure their own metrics and success.

**Previous Work on the Topic**

Most of the analysis done around Twitch.tv’s data is in regards to the platform as a whole. Students from WPI found relationships between the demographics and the viewership behavior in relation to the types and formats of the content being delivered (Farrington and Muesch). Other breakdowns look at aggregate metrics for Twitch, such as the numbers of broadcasters and peak viewership in order to generate a tentative estimate of Twitch’s revenue (Iqbal). Even recommendations for marketing managers and potential influencers are presented with aggregate data metrics (Influencer Marketing Hub). Given the background of Twitch as a relatively new platform, we were unable to find any analysis on specific channels or the application of machine learning methods to predict singular channel growth. Identifying this gap and realizing its potential, we set out to create models that would allow us to explore Twitch through the lens of singular channels and learn from applying ML methods to real data.

**Questions/Hypotheses**
* **Research Questions**
    * Which features are most important in predicting the expected future income growth of Twitch channels?
    * What is the ideal number of features to include in analysis when predicting income growth of Twitch channels?
    * Which supervised machine learning algorithm will produce the highest accuracy for our test dataset?

* **Hypotheses 1**
    * Hypotheses: Followers gained is the most influential feature in predicting income growth. 
    * Null: Followers gained has is not the most influential feature in predicting income growth. 

* **Hypotheses 2**
    * Hypotheses: Stream time has a positive relationship with income growth. 
    * Null: Stream time has no relationship with income growth. 


* **Hypotheses 3**
    * Hypotheses: A significant relationship exists between view gained and income growth. 
    * Null: No relationship exists between views gained and income growth.


**References**

Iqbal, Mansoor. “Twitch Revenue and Usage Statistics (2020).” Business of Apps, Sokomedia, 24 Apr. 2020, www.businessofapps.com/data/twitch-statistics/

“25 Useful Twitch Stats for Influencer Marketing Managers [Infographic].” Influencer Marketing Hub, 10 June 2020, influencermarketinghub.com/twitch-stats/

Bellanger, Caroline. “New Study from Upfluence Finds COVID-19 Lockdown Restrictions Resulted in a 24% Viewership Increase on Live-Streaming Platform Twitch.” Upfluence, Upfluence, 3 Apr. 2020, www.upfluence.com/press-release/new-study-from-upfluence-finds-covid-19-lockdown-restrictions-resulted-in-a-24-viewership-increase-on-live-streaming-platform-twitch

Farrington, Daniel Jeffrey, and Muesch, Nicholas Matthew. “Analysis of the Characteristics and Content of Twitch Live-Streaming.” Digital WPI, Worcester Polytechnic Institute, Mar. 2015, digitalcommons.wpi.edu/iqp-all/3376/


<a id="2"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 2. METHOD

### 2.1. Data Acquisition

We sourced the data we are using ourselves through web-scraping two websites, TwitchTracker and SullyGnome, where both sites track and record channel metrics from Twitch.

From **SullyGnome**, we were able to obtain data of the 100 most popular Twitch streamers per month. The data included information such as watch time, stream time, peak viewers, etc. and we sourced data for 13 months, ranging from April 2019 to April 2020. We obtained a total number of 11 variables from the dataset.

We took subscriber data from **TwitchTracker** and used it to generate the channel’s income data, then appended the income data to a cleaned aggregate dataset from SullyGnome. This variable is the outcome variable. 

Twitch: https://twitchtracker.com/subscribers
SullyGnome: https://sullygnome.com 

### 2.2. Variables

**Feature Variables:**
* Watch time
* Stream Time
* Peak Viewers
* Average Viewers
* Followers
* Followers Gained
* Views Gained
* Partnered
* Mature
* Language

**Target Variable:**
* Income Growth

For our analysis purposes, we used the feature variables as independent variables and the target variable as the dependent variable for all our hypotheses. 



### 2.3. Data Analysis
**Feature Extraction:** We will use Iterative Feature Selection to gain an understanding of which variables are important and the order of importance of the variables. This will allow us to gain more insight into our data. For analysis, we will use Model-Based Feature Selection to judge which features are most important for use in our model. 

**Machine Learning Techniques:** We will be applying supervised machine learning algorithms to analyze the data. As our target variable of future income is a continuous variable, we plan on using regression analysis on our feature variables. We will be applying Multiple Linear Regression, Lasso, Ridge, KNNRegressor and SVM machine learning algorithms to our data to analyze for best results as they are all used to tune regression models. We expect Lasso to perform the best as our data contains many features, but we expect only a few of them to be impactful to the target variable. 

<a id="3"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 3. RESULTS

### 3.1. Data Wrangling


**Simple Data Cleaning:**

For the data obtained from SullyGnome, we merged all of the files from each month together. In order to be able to sufficiently train our machine for each user, we decided that we need to obtain at least 10 data points per channel so we omitted all users who did not appear in the dataset at least 10 times. Next, we adjusted the channel names as, in some instances, the foreign name and the English name of the channel would appear in the data. We cleaned it so that only the English name appeared, which would then be used by the subscriber scraper. 

For the subscriber scraper, we scraped data from the Twitch site for each user. The scraping takes about 2-3 hours. From that information, we calculated the current subscriber based income of the users by performing basic multiplication based on the subscriber count per tier. 

The final dataset will merge the total income variable from the Twitch data and the data from SullyGnome. The variables will merge together on the accounts of channel name and data. 


In [102]:
import pandas as pd
import numpy as np

allchannels = pd.read_csv("generated_data/TotalChannelData.csv")
allchannels.set_index("Channel", inplace=True)

subdata = pd.read_csv("generated_data/SubData.csv")
subdata.set_index("Channel", inplace=True)

finaldata = pd.merge(allchannels, subdata, how='left', on=['Channel', 'Month'])
finaldata.drop(['Unnamed: 0', "Prime", "Tier 1", "Tier 2", "Tier 3", "Total", "Unshared"], axis =1, inplace=True)

#removing all rows where total income is not provided, meaning subscriber data was not scraped 
#this means the rows being removed are all channels that have less than 10 datapoints
finaldata = finaldata[finaldata['Total Income'].notna()]

#removing dollar formatting from total income 
finaldata['Total Income'] = finaldata['Total Income'].str.replace(',', '').str.replace('$', '').astype(float)

In [103]:
#check if there are any partnered channels that are "False"
def check_continuous_variables(df, col):
    false_count = 0
    for i in range(len(finaldata.loc[:,col])):
        if finaldata.loc[:,col][i] == False:
            false_count +=1 
        else:
            continue
    print("There are", false_count, "false values for", col)
    if false_count == 0:
        finaldata.drop([col], axis = 1, inplace=True)
    return finaldata

In [104]:
finaldata = check_continuous_variables(finaldata, "Partnered")

There are 0 false values for Partnered


In [105]:
finaldata = check_continuous_variables(finaldata, "Mature")

There are 377 false values for Mature


Within the final dataset, we needed to calculate income growth. To analyze income growth, we needed to get the total income of the first month of data and subtract it from the next month of data. That was done per channel for the duration of the time period of data we had and merged back together with the main dataset. 

In [106]:
finaldata = finaldata.sort_values(by=["Channel", "Month"], ascending=True)

channels = finaldata.index.unique()
i = 0 
IncomeChange_list = []
for channel in channels:
    channeldata = finaldata[finaldata.index == channel]
    prev = 0
    for index, row in channeldata.iterrows():
        if(prev != 0):
            row["Income Change"] = row["Total Income"] - prev
            IncomeChange_list.append([channel, row["Income Change"], row["Month"]])
        else:
            prev = row["Total Income"]
            IncomeChange_list.append([channel, np.nan , row["Month"]])

IncomeChange = pd.DataFrame(IncomeChange_list, columns = ["Channel", "Income Change", "Month"])
IncomeChange = IncomeChange.set_index("Channel")


finaldata = pd.merge(finaldata, IncomeChange, on = ["Channel", "Month"])
            
finaldata = finaldata[finaldata['Income Change'].notna()]

finaldata.drop(columns=["Total Income"], axis = 1, inplace = True)

finaldata.head()

Watch time  Stream time  Peak viewers  Average viewers  \
Channel                                                                 
AdmiralBahroo   110374845        13920         14616             7928   
AdmiralBahroo    84227100        11910         12756             7071   
AdmiralBahroo   110024835        15015         29431             7327   
AdmiralBahroo    74971200         9870         15925             7595   
AdmiralBahroo    75818505        10530         54643             7199   

               Followers  Followers gained  Views gained  Mature Language  \
Channel                                                                     
AdmiralBahroo     770090              9819       1423087   False  English   
AdmiralBahroo     770090              8651       1229356   False  English   
AdmiralBahroo     770090             14489       1376799   False  English   
AdmiralBahroo     767646             11515       1045862   False  English   
AdmiralBahroo     767646             15848       1157761   False  English   

                    Month  Income Change  
Channel                                   
AdmiralBahroo  2019-01-01      -36295.11  
AdmiralBahroo  2019-02-01       -7912.10  
AdmiralBahroo  2019-03-01      -21266.30  
AdmiralBahroo  2019-04-01      -57068.87  
AdmiralBahroo  2019-05-01      -24214.50

#### Identifying feature and target variables

In [107]:
#function returns features and target variables 
def features_target(df, target_var):
    features = df.drop(target_var, axis = 1)
    features = features.reset_index() #index is reset to merge with encoded features 
    target = df[target_var]
   
    return features, target


In [108]:
features, target = features_target(finaldata, "Income Change")

In [109]:
features.head()

Channel  Watch time  Stream time  Peak viewers  Average viewers  \
0  AdmiralBahroo   110374845        13920         14616             7928   
1  AdmiralBahroo    84227100        11910         12756             7071   
2  AdmiralBahroo   110024835        15015         29431             7327   
3  AdmiralBahroo    74971200         9870         15925             7595   
4  AdmiralBahroo    75818505        10530         54643             7199   

   Followers  Followers gained  Views gained  Mature Language       Month  
0     770090              9819       1423087   False  English  2019-01-01  
1     770090              8651       1229356   False  English  2019-02-01  
2     770090             14489       1376799   False  English  2019-03-01  
3     767646             11515       1045862   False  English  2019-04-01  
4     767646             15848       1157761   False  English  2019-05-01

In [110]:
target.head()

Channel
AdmiralBahroo   -36295.11
AdmiralBahroo    -7912.10
AdmiralBahroo   -21266.30
AdmiralBahroo   -57068.87
AdmiralBahroo   -24214.50
Name: Income Change, dtype: float64

#### Encoding continuous feature variables 

In [111]:
from sklearn.preprocessing import OneHotEncoder

#one-hot encoding for "Mature" and "Language" Variables 
encode_features = features.iloc[:, 8:10]

encoder = OneHotEncoder(sparse = False)
encoded_df = encoder.fit_transform(encode_features)

features_df = pd.DataFrame(encoded_df, columns = encoder.get_feature_names())

#merging both features datasets together 
features = features_df.merge(features, left_index = True, right_index = True)
features.drop(["Mature", "Language"], axis = 1, inplace=True)

#removing channel and month as a variables as they work as identifiers 
features.reset_index()
features.drop(["Channel", "Month"], axis = 1, inplace=True)

In [112]:
featuresheading = features.columns
features.head()

x0_False  x0_True  x1_English  x1_French  x1_German  x1_Korean  \
0       1.0      0.0         1.0        0.0        0.0        0.0   
1       1.0      0.0         1.0        0.0        0.0        0.0   
2       1.0      0.0         1.0        0.0        0.0        0.0   
3       1.0      0.0         1.0        0.0        0.0        0.0   
4       1.0      0.0         1.0        0.0        0.0        0.0   

   x1_Portuguese  Watch time  Stream time  Peak viewers  Average viewers  \
0            0.0   110374845        13920         14616             7928   
1            0.0    84227100        11910         12756             7071   
2            0.0   110024835        15015         29431             7327   
3            0.0    74971200         9870         15925             7595   
4            0.0    75818505        10530         54643             7199   

   Followers  Followers gained  Views gained  
0     770090              9819       1423087  
1     770090              8651       1229356  
2     770090             14489       1376799  
3     767646             11515       1045862  
4     767646             15848       1157761

#### Scaling our feature variables

In [113]:
from sklearn.preprocessing import MinMaxScaler

def scale_the_dataset():          
    #create the scaler
    scaler = MinMaxScaler()
    #fit scaler to training data
    scaler.fit(features) 
    scaled_features = scaler.transform(features)     
    
    return scaled_features

In [114]:
featuresarray = scale_the_dataset()

features = pd.DataFrame(featuresarray, columns = featuresheading)
features

x0_False  x0_True  x1_English  x1_French  x1_German  x1_Korean  \
0         1.0      0.0         1.0        0.0        0.0        0.0   
1         1.0      0.0         1.0        0.0        0.0        0.0   
2         1.0      0.0         1.0        0.0        0.0        0.0   
3         1.0      0.0         1.0        0.0        0.0        0.0   
4         1.0      0.0         1.0        0.0        0.0        0.0   
..        ...      ...         ...        ...        ...        ...   
498       1.0      0.0         1.0        0.0        0.0        0.0   
499       1.0      0.0         1.0        0.0        0.0        0.0   
500       1.0      0.0         1.0        0.0        0.0        0.0   
501       1.0      0.0         1.0        0.0        0.0        0.0   
502       1.0      0.0         1.0        0.0        0.0        0.0   

     x1_Portuguese  Watch time  Stream time  Peak viewers  Average viewers  \
0              0.0    0.024288     0.267896      0.018596         0.051686   
1              0.0    0.011643     0.219450      0.012437         0.042965   
2              0.0    0.024118     0.294288      0.067653         0.045570   
3              0.0    0.007167     0.170282      0.022931         0.048297   
4              0.0    0.007576     0.186189      0.151137         0.044268   
..             ...         ...          ...           ...              ...   
498            0.0    0.153758     0.376717      0.115845         0.179716   
499            0.0    0.143591     0.375633      0.071875         0.168593   
500            0.0    0.151117     0.330803      0.100497         0.200405   
501            0.0    0.547265     0.438539      0.707687         0.548542   
502            0.0    0.240034     0.355025      0.164852         0.293967   

     Followers  Followers gained  Views gained  
0     0.059247          0.042237      0.039939  
1     0.059247          0.040969      0.032841  
2     0.059247          0.047307      0.038243  
3     0.058949          0.044078      0.026119  
4     0.058949          0.048783      0.030218  
..         ...               ...           ...  
498   0.330638          0.142878      0.201807  
499   0.330638          0.144371      0.185142  
500   0.330638          0.274375      0.211780  
501   0.330638          0.682716      0.561239  
502   0.336338          0.120850      0.312777  

[503 rows x 14 columns]

#### Splitting data into training and testing data

In [115]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

#split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features, target, random_state=3000)

#### Iterative Feature Selection 
We ran iterative feature selection, with the number of features ranging from 1 to the total number of features in order to develop a list of all features, order from most important to least important. 


In [116]:
#Running Iterative Feature Selection to return a list of features in order of importance 
from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeRegressor


def ife_features(X_train, y_train, X_test, y_test):
    name_features = list(X_train.columns.values)
    num_features = len(name_features)
    features_names = []
    
    for i in range(num_features):
        i += 1
        select = RFE(DecisionTreeRegressor(random_state = 3000), n_features_to_select = i)
        
        #fit the RFE selector to the training data
        select.fit(X_train, y_train)
        
        selected_features = select.get_support()
        
        #adding to the list of features in order of appearance / importance 
        for x in range(num_features):
            if selected_features[x] == True:
                if name_features[x] in features_names:
                    continue
                else:
                    features_names.append(name_features[x])
                    
    print("Order of Importance of Features:")
    order = 1
    for value in features_names:
        print("\t",order,value)
        order+= 1


In [117]:
ifefeatures = ife_features(X_train, y_train, X_test, y_test)
ifefeatures

Order of Importance of Features:
	 1 Followers
	 2 Views gained
	 3 Followers gained
	 4 Watch time
	 5 x0_False
	 6 Peak viewers
	 7 Average viewers
	 8 Stream time
	 9 x0_True
	 10 x1_English
	 11 x1_German
	 12 x1_Portuguese
	 13 x1_French
	 14 x1_Korean


#### Model Based Feature Selection 
As there was no way to determine a defined number of features to use for the data, model based feature selection was used to develop the list of selected features. This feature selection model gave a list of 5 total features to act as the selected criteria. 

In [118]:
from sklearn.feature_selection import SelectFromModel
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

def modelbased_features(X_train, y_train, X_test, y_test):
    
    name_features = list(X_train.columns.values)
    
    select = SelectFromModel(DecisionTreeRegressor(random_state = 3000))

    select.fit(X_train, y_train)
    X_train_selected = select.transform(X_train)
    X_test_selected = select.transform(X_test)

    selected_features = select.get_support()
    feature_names = []
    for x in range(len(selected_features)):
        if selected_features[x] == True:
            feature_names.append(name_features[x])
            
    return feature_names, X_train_selected, X_test_selected

In [119]:
modelbasedfeatures, X_train_selected, X_test_selected = modelbased_features(X_train, y_train, X_test, y_test)
modelbasedfeatures

['x0_True', 'Watch time', 'Followers', 'Followers gained', 'Views gained']

### 3.2. Data Exploration

In [120]:
import plotly.express as px
import plotly.graph_objects as go

The scatter plot below shows followers versus income gained. Because our dataset has infrequent updates for the followers count (from the data that we sourced) there is no correlation whatsoever between follower count and change in income as reflected from the data.

In [121]:
fig = px.scatter(finaldata, x="Followers", y="Income Change", trendline="ols")
fig.show()

<img src="https://raw.githubusercontent.com/Liyih/Twitch.tv-Metrics-Analysis/master/img/FollowersVsIncome.png">

The Histogram below shows the distribution of monthly income changes for our top 50 streamers. As the bell curve indicates, the income variation from month to month remains largely unchanged as a whole - and large fluctuations in income are rare.

In [122]:
fig = px.histogram(finaldata, x="Income Change", nbins=50)
fig.show()

<img src="https://raw.githubusercontent.com/Liyih/Twitch.tv-Metrics-Analysis/master/img/IncomeHisto.png">

The violin plot below tracks the volatility of income levels on a monthly basis - it is clear that post COVID-19 (around March 2020) there was a resultant increase in subscription activity across all major channels, with a slightly positive median (more channels benefitted from COVID than were hurt). 

In [123]:
fig = px.violin(finaldata, y="Income Change", x="Month")
fig.show()

<img src="https://raw.githubusercontent.com/Liyih/Twitch.tv-Metrics-Analysis/master/img/IncomeViolin.png">

The scatter plot below tracks followers gained versus income gained. This graph indicates a positive relationship between gaining followers and gaining income, which follows the logic that an increase in followers could potentially lead to an increase in subscribers (income).

In [124]:
fig = px.scatter(finaldata, x="Followers gained", y="Income Change", title="Followers gained vs Income Change", trendline="ols")
fig.show()

<img src="https://raw.githubusercontent.com/Liyih/Twitch.tv-Metrics-Analysis/master/img/FollowersGainedVsIncome.png">

The scatter plot below tracks watch time versus income gained. Despite the visual indication of the sloped line, the actual slope of the line is ~0.00004 and has no discernible correlation between the effect of watch time on income. 

In [125]:
fig = px.scatter(finaldata, x="Watch time", y="Income Change", title="Watch Time vs Income Change", trendline="ols")
fig.show()

<img src="https://raw.githubusercontent.com/Liyih/Twitch.tv-Metrics-Analysis/master/img/WatchTimeVsIncome.png">

Similarly to the graph above, there is little to no correlation between streamer time to the streamer’s income; therefore streaming more or having a larger audience does not guarantee a significant level of change in their subscriber (income) levels.

In [134]:
fig = px.scatter(finaldata, x="Stream time", y="Income Change", title="Stream Time vs Income Change", trendline="ols")
fig.show()

<img src="https://raw.githubusercontent.com/Liyih/Twitch.tv-Metrics-Analysis/master/img/StreamTimeVsIncome.png">

### 3.3. Model Construction

We normalize our data then conduct oneway ANOVA tests on our feature and target pairs.

In [127]:
from scipy import stats
from statsmodels.stats.multicomp import MultiComparison
from statsmodels.stats.multicomp import pairwise_tukeyhsd
from sklearn.preprocessing import normalize 

tonorm_data = finaldata.drop(columns = ["Language", "Month"], axis =1)
norm_data = normalize(tonorm_data)

norm_data = pd.DataFrame(norm_data, columns = tonorm_data.columns)

def oneway_ANOVA(data, IV, DV):
    oneway = stats.f_oneway(data[IV], data[DV])
    fstat = oneway[0]
    pval = oneway[1]
    df1 = len(data.groupby(IV)) - 1
    df2 = len(data.groupby(DV)) - 1
    print("\n\nF-test:", IV, "and", DV)
    print("-------------------------------------------\n")
    print("F(%d,%d) = %.2f, p = %s" %(df1, df2, fstat, pval))
    print("\nAssumption of Normality:")
    shapiro_test_IV = stats.shapiro(data[IV])
    print("\t" + IV + ": " + str(shapiro_test_IV))
    print("\t\t", end="")
    checkAssumption(shapiro_test_IV[1])
    shapiro_test_DV = stats.shapiro(data[DV])
    print("\t" + DV + ": " + str(shapiro_test_DV))
    print("\t\t", end="")
    checkAssumption(shapiro_test_DV[1])
    
#Print the correct assumption statement
def checkAssumption(pval):
    if pval > 0.05:
        print("Assumption is met. p > .05")
    else:
        print("Assumption is violated. p < .05")

oneway_ANOVA(norm_data, "Followers", "Income Change")
oneway_ANOVA(norm_data, "Peak viewers", "Income Change")
oneway_ANOVA(norm_data, "Views gained", "Income Change")
oneway_ANOVA(norm_data, "Followers gained", "Income Change")



F-test: Followers and Income Change
-------------------------------------------

F(502,502) = 898.57, p = 1.6019400153533745e-141

Assumption of Normality:
	Followers: (0.8745790719985962, 1.0206135030008e-19)
		Assumption is violated. p < .05
	Income Change: (0.7892618775367737, 5.400846454667688e-25)
		Assumption is violated. p < .05


F-test: Peak viewers and Income Change
-------------------------------------------

F(502,502) = 907.92, p = 1.365399948983186e-142

Assumption of Normality:
	Peak viewers: (0.7437337636947632, 3.71416437184254e-27)
		Assumption is violated. p < .05
	Income Change: (0.7892618775367737, 5.400846454667688e-25)
		Assumption is violated. p < .05


F-test: Views gained and Income Change
-------------------------------------------

F(502,502) = 3597.00, p = 0.0

Assumption of Normality:
	Views gained: (0.9158664345741272, 4.1062494980729494e-16)
		Assumption is violated. p < .05
	Income Change: (0.7892618775367737, 5.400846454667688e-25)
		Assumption is vi

In [128]:
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import LinearSVR
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

In [129]:
estimators = {"Linear Regression": LinearRegression(), "Ridge": Ridge(), "Lasso":Lasso(), "k-Nearest Neighbor":
             KNeighborsRegressor(), "Support Vector Machine": LinearSVR(max_iter=10000000)}

In [130]:
def regressors_percentage_split(X_train, X_test, y_train, y_test):
    for value in estimators:
        model = estimators[value].fit(X=X_train, y=y_train)
        r_train = r2_score(y_train, model.predict(X_train))
        r_test = r2_score(y_test, model.predict(X_test))
        print(value+":")
        print("\tR-squared value for training set: ",r_train)
        print("\tR-squared value for testing set: ",r_test,"\n" )
    return



In [131]:
#creating test and train datasets for when followers is not included in features

features_nofollower = features.drop(columns=["Followers"], axis =1)
X_train_nofollower, X_test_nofollower, y_train_nofollower, y_test_nofollower = train_test_split(
    features_nofollower, target, random_state=3000)
nofollower_modelbasedfeatures, X_train_selected_nofollower, X_test_selected_nofollower = modelbased_features(
    X_train_nofollower, y_train_nofollower, X_test_nofollower, y_test_nofollower)

In [132]:
print("Analyzing Using All Features")
print("-------------------------------------------\n")
print("When including followers")
regressors_percentage_split(X_train, X_test, y_train, y_test)
print("When excluding followers")
regressors_percentage_split(X_train_nofollower, X_test_nofollower, y_train_nofollower, y_test_nofollower)

print("Analyzing Using Selected Features")
print("-------------------------------------------\n")
print("When including followers")
regressors_percentage_split(X_train_selected, X_test_selected, y_train, y_test)
print("When excluding followers")
regressors_percentage_split(X_train_selected_nofollower, X_test_selected_nofollower, y_train_nofollower, 
                            y_test_nofollower)



Analyzing Using All Features
-------------------------------------------

When including followers
Linear Regression:
	R-squared value for training set:  0.18683833606044442
	R-squared value for testing set:  0.35477624162092514 

Ridge:
	R-squared value for training set:  0.16501107000093984
	R-squared value for testing set:  0.31890015393104065 

Lasso:
	R-squared value for training set:  0.18679367660084745
	R-squared value for testing set:  0.35427804518218675 

k-Nearest Neighbor:
	R-squared value for training set:  0.7069041247383286
	R-squared value for testing set:  0.448037864702433 

Support Vector Machine:
	R-squared value for training set:  -0.0016805319004558061
	R-squared value for testing set:  -0.0005683852052307259 

When excluding followers
Linear Regression:
	R-squared value for training set:  0.12251164627431521
	R-squared value for testing set:  0.33107280807624206 

Ridge:
	R-squared value for training set:  0.10444268929228973
	R-squared value for testing set:  0

### 3.4. Model Evaluation

#### Machine Learning
* Five machine learning algorithms were analyzed: Multiple Linear Regression, Ridge, Lasso, k-Nearest Neighbor, and Support Vector Machine. They were analyzed in four different situations: analyzing all features including followers, all features excluding followers, selected features including followers, and selected features excluding followers.
* Among all conditions, Support Vector Machine performed the worst by having negative R-squared values for all condition. 
* Multiple Linear Regression, Ridge, and Lasso had the similar results, but did consistently worse than k-Nearest Neighbors. 
* k-Nearest Neighbors performed the best out of all 5 algorithms. Among the situations, k-Nearest Neighbors performed better when including "Followers", so when we performed tuning, we focused on only two situations of analysis where we looked at all features vs. selected features. The best performance occured on selected features, when including "Followers". In that situation, the R-squared for training data was .734 and for testing data was .402. For all features including "Followers", the R-squared had a .03 difference, landing at .707 for training and .448 for testing data. This is at risk for overfitting as the R-sqaured for training data is higher than that for testing data.

#### Questions
* The most important feature identified in predicting future income growth of Twitch channels via model based feature selection was Followers. However, when graphing the regression line, it resulted in a slope of 0.0002. In addition, we believed the data for Followers was corrupt in the dataset. So, to account for this, we also performed machine learning algorithms on features selections that excluded “Followers”.
* From the model based feature selection, it was found that the ideal number of features to include in analysis was 5 features. However, when performing machine learning algorithms, higher R-squared was achieved when analyzing with all features rather than with the selected features. 
* The supervised machine learning algorithms that produced the highest accuracy were Multiple Linear Regression, Ridge Regression, and Lasso Regression. 

#### Hypothesis Testing
* **Hypotheses 1:** Followers gained was not the most influential feature in predicting income growth. Through iterative feature selection, the list of features in order of importance of influence to the target variable of income growth were shown. Followers was considered most important, as it was at the top of the list. Thus, hypotheses 1 is rejected and it can be concluded that followers gained is not the most influential feature in predicting income growth. 
* **Hypotheses 2:** The regression line on the scatter plot comparing stream time and income change is shown to have the equation of approximately y =.096x -1567. This shows a small but positive relationship between stream time and income change. However, the ANOVA shows that the p value of the relationship does not pass the assumption of normality as it has a p value of less than 0.05, thus no significant positive relationship exists. Thus, hypothesis 2 is rejected and it can be concluded that stream time has no relationship with income growth. 
* **Hypothesis 3:** Through ANOVA analysis of views gained and income growth, hypothesis 3 is rejected as the data does not pass the assumption of normality because the p value is less than 0.05. Thus, it can be concluded that no relationship exists between views gained and income growth


### 3.5. Model Optimization & 3.6. Model Testing

In [133]:
from sklearn.model_selection import GridSearchCV

import warnings
warnings.filterwarnings('ignore')

param_grid = {"n_neighbors":[1, 5, 10], "metric": ["euclidean", "manhattan", "minkowski"]}

def grid_search_kNN():
    grid_search = GridSearchCV(KNeighborsRegressor(), param_grid, cv=5)
    grid_search.fit(X=X_train, y=y_train)
    
    print("Non-scaled:")
    print("Best parameters: ", grid_search.best_params_)
    print("Training set score with best parameters: ", grid_search.score(X_train, y_train))
    print("Test set score with best parameters: ", grid_search.score(X_test, y_test))

grid_search_kNN()

print("\n")

def grid_search_selected_kNN():
    grid_search = GridSearchCV(KNeighborsRegressor(), param_grid, cv=5)
    grid_search.fit(X=X_train_selected, y=y_train)
    
    print("Scaled:")
    print("Best parameters: ", grid_search.best_params_)
    print("Training set score with best parameters: ", grid_search.score(X_train_selected, y_train))
    print("Test set score with best parameters: ", grid_search.score(X_test_selected, y_test))

grid_search_selected_kNN()

Non-scaled:
Best parameters:  {'metric': 'manhattan', 'n_neighbors': 5}
Training set score with best parameters:  0.7253479389200777
Test set score with best parameters:  0.5018932742254127


Scaled:
Best parameters:  {'metric': 'manhattan', 'n_neighbors': 5}
Training set score with best parameters:  0.7452388820640092
Test set score with best parameters:  0.42321741043859984


We used GridSearch to tune kNN specifically and try to improve its performance accuracy by attempting to correct the overfitting. However, after running GridSearch, while we did see an improvement in our non-selected features (from .42 to .50 accuracy on the testing set), that performance improvement was not significant enough to correct our model’s overfitting. Additionally, despite the performance improvement, the overall accuracy of our model is too low to make confident predictions on new data. Given this result, we conclude that a model with high prediction accuracy is unable to fit onto our data, and that it is difficult to tune our data’s features within our dataset’s limits to accurately predict our target. 

<a id="4"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 4. DISCUSSION
* Provide a summary of the steps you took to analyze your data and test your predictive model
* Interpret your findings from 3.4., 3.5, and 3.6
    * Which algorithms did you compare?
    * Which algorithm(s) revealed best performance?
    * Which algorithm(s) should be used for your predictive model?
    * Based on your findings, can we you the features in your dataset to predict the outcome variable you identified using the algorithms you've applied? (It is okay if the answer is no. We're interested in the process, not the performance of the model.)
* For your hypotheses, interpret the results. What does it mean to have significant/non-significant differences with regards to your data?


* End this section with a conclusion paragraph containing some pointers for future work
    * (e.g., get more data/features, perform another analysis, etc.)

<a id="5"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

### CONTRIBUTIONS
* Describe each team member's contributions to the report (who did what in each section)
* Remember this is a team effort!
* Each member of your team will provide peer evaluation of other team members. Your final grade on the project will be based on those peer evaluations. An survey will be shared after the deadline for this deliverable.